feat(core): support server set cookies

MathiasCiarlo opened the pull request:
This fixes the issue that arises when trying to access a protected path directly(the first request to the website). The middleware tries to save the attempted, protected path to storage, so that we can redirect to the path after login. The problem was that the "auth.redirect" cookie was not set when the middleware was run server side. This PR supports cookies set server side and should fix #332, fixes #326 and fixes #134.

karimcambridge commented:
I also encountered this issue!

mhafizhasan commented:
+1 should be fixed

sambrezo commented:
This is an elegant solution to the problem I am having with users getting logged out because the token has changed server side but is not synced with the client.

sambrezo commented:
I have found two potential problems with this PR:

1. getCookie(setCookie()) is not working as expected
  Breaks the Storage class contract
2. setCookie(A) && setCookie(B) overwrites A
  Can be fixed by more careful handling of `Set-Cookie` headers
3. Cookie options are ignored

I need to pass several cookies back to the client, so I am thinking of setting them in nuxtServerInit instead.

MathiasCiarlo commented:
> I have found some potential problems with this PR:
> 
> 1. getCookie(setCookie()) is not working as expected
>    Breaks the Storage class contract
> 2. setCookie(A) && setCookie(B) overwrites A
>    Can be fixed by more careful handling of `Set-Cookie` headers
> 3. Cookie options are ignored
> 
> I need to pass several cookies back to the client, so I am thinking of setting them in nuxtServerInit instead.

Thank you for the feedback! I have fixed the overwriting issue, and added options. What do you mean by with `getCookie(setCookie())` ? The setCookie function still returns the set value?

sambrezo commented:
Hey @MathiasCiarlo,
that's looking good. I'm gonna test it with the rest of my app this week and let you know if it works for me. Thanks for tackeling the cookie issue!

Regarding 1) I think the issue might still persist. The problem that I see is when you call `setCookie` while SSR and then read the value again with `getCookie` you will get the cookie information from the request header and not what you have just written to the response header. Might be an edge case but all other getters of the Storage class provide some kind of persistence which is not guaranteed when he have to deal with two sets of cookies. No 3) is somewhat related in that the cookies should stay in sync when moving server to client and vice-versa. Unfortunately the cookie.options in `auth-module`do not always map onto the values we see in the cookie string. For example `expires` is used differently in `js-cookie`.
We also have to keep in mind that we cannot use HTTP enabled cookies cookies server-side because we rely on the client being able to read and write to the cookie via `js-cookie`. This means we cannot really protect our tokens from XSS when we move forward with this approach.

Would be nice to have one of the other contributors comment on the decision whether to support server cookies at all.

zachzoup commented:
Is this issue crashing the tab when you navigate to localhost:3000 (page that requires auth)?

duyleekun commented:
I `patch-package` this to use this fix immediately. Hope this get merged soon

rchl commented on lib/core/storage.js:
```diff
@@ -179,9 +179,31 @@ export default class Storage {
+        if (!setCookieHeader) {
+          setCookieHeader = []
+        }
+        const existingCookieIndex = setCookieHeader.findIndex(element => element.split('=')[0] === _key)
```
I would feel more confident using dedicated package for parsing cookies like `cookie`. Also it would make code a lot cleaner.

See example in `nuxt-i18n` module on handling server side cookie using that module:
https://github.com/nuxt-community/nuxt:
 - set: i18n/blob/ca9873b85f0d1e69763cafca01a7c7df51893107/src/plugins/main.js#L63,L74
 - read: https://github.com/nuxt-community/nuxt-i18n/blob/ca9873b85f0d1e69763cafca01a7c7df51893107/src/templates/middleware.js#L43,L48

MathiasCiarlo replied:
Totally agree! Will fix it :)

pi0 replied:
Agree with @rchl. `cookie` has ~1KB overhead but seems safer. We can require it inside `if (process.server)` to prevent including in client-side bundle. Please ping me when you refactored to get merged ASAP.

rchl replied:
Actually this file even imports those packages already:

```js
import Cookies from 'js-cookie'
...
import { parse as parseCookie } from 'cookie'
```

rchl commented on lib/core/storage.js:
```diff
@@ -179,9 +179,31 @@ export default class Storage {
     if (isUnset(value)) {
       Cookies.remove(_key, _options)
```
I haven't looked or understood the logic of the module and call-sites using this code but this code should probably also have server-side branch to remove cookie (regardless if this code is used server-side or not).

MathiasCiarlo replied:
I think it's being used after fetching the url to redirect back to the original page after a successful login. When the redirect cookie is fetched, it is set to null. So this will not work ssr, unless we add the branch you are talking about. Fixing it now :)

rchl commented on lib/core/storage.js:
```diff
@@ -179,9 +179,24 @@ export default class Storage {
+        cookies[_key] = value
+
+        // Convert back to array and set the header
+        const newCookieArray = Object.keys(cookies).map(key => serializeCookie(key, cookies[key]))
```
`serializeCookie` is not imported

MathiasCiarlo replied:
Oops, should be ok now :) Now I only need to look at the isUnset thing above.

rchl commented on lib/core/storage.js:
```diff
@@ -179,9 +179,25 @@ export default class Storage {
 
+      // Support server set cookies
+      if (process.server) {
+        const { parse, serialize } = require('cookie')
```
It doesn't make much sense to import only in this branch if it's also imported at the top of the file.

Also, you shouldn't ideally use `require` in a file that is using es modules syntax.

It's probably OK to always (regardless if server or not) import IMO.

MathiasCiarlo replied:
I also think importing at the top is best. Is the performance boost really worth it? Or should I refactor to only using the already existing import at the top? @pi0 
```
import { parse as parseCookie, serialize as serializeCookie } from 'cookie'
```

pi0 replied:
Counting tree-shaking it saves `serialize` from client bundle. But agree about code readability to move to top import.

MathiasCiarlo commented:
I've added support for unsetting ssr cookies. The only thing remaining is to think about cookie options [link to docs](https://auth.nuxtjs.org/v/dev/getting-started/options#cookie). In an earlier [commit](https://github.com/nuxt-community/auth-module/commit/66433cff7d173c3bbbbd8c343ce122978e1b170a), I did a simple implementation, but @sambrezo told me the options used in js-cookie are a little different than my implementation. Do we have to include options ssr, and if so, please help me out in how we can serialize js-cookie options with the current implementation :)

pi0 commented:
Hi @MathiasCiarlo. I've proposed a PR to backport universal-storage improved storage into auth module. Would you please have a review on it?

I think the difference with this PR is that here we support multiple `Set-Header` calls. If you think it is a requirement we can combine these PRs and use `getCookiesFromResponseHeader`/`setResponseHeaderCookies` utils.

MathiasCiarlo commented:
Great @pi0, I'll have a look tonight :)

MathiasCiarlo commented:
@pi0 Have looked at your PR. Many nice changes, but we definitely need the multiple Set-Header calls. How can we combine the PRs? Can I raise a new PR, branched out from your branch? :)

pi0 commented:
Thanks, everyone for patientnce and helping on this issue. Sever set cookie is added in [v4.6.0](https://github.com/nuxt-community/auth-module/releases/tag/v4.6.0) via #360.
